File: 003 - Largest prime factor/Haskell/pe_003.hs
{- |
file:   pe_003.hs
title:  Largest prime factor
author: Alexander P. Gosselin
e-mail: alexandergosselin@gmail.com
date:   September 20, 2014
-}

largestPrimeFactor :: Integer -> Integer -> Integer
largestPrimeFactor x y
    | x == y        = x
    | mod x y == 0  = largestPrimeFactor (quot x y) y
    | otherwise     = largestPrimeFactor x (y + 2)

pe_003 = largestPrimeFactor 600851475143 3

main :: IO ()
main = do
    print pe_003

File: 003 - Largest prime factor/Fortran/pe_003.f90
! file:   pe_003.f90
! title:  Largest prime factor
! author: Alexander Gosselin
! e-mail: alexandergosselin@gmail.com
! date:   May 8, 2015
! 
! link:   https://projecteuler.net/problem=1
! 
! copyright: (C) 2015 Alexander Gosselin
! license:   GNU General Public License <http://www.gnu.org/licenses/>

program pe_003
  implicit none 
  integer, parameter :: int64 = selected_int_kind(13)
  integer(int64) :: n = 600851475143_int64
  integer :: i = 3
  do while (i*i < n)
    if (mod(n, i) == 0) then
      n = n/i
    else
      i = i + 2
    end if
  end do
  print '(I0)', n
end program
